### ISSUE TITLE
Template generics as parameters/response types

### ISSUE THREAD

**Issue by tjprescott** [microsoft/cadl#551]
While working on TextAnalytics, I had something along these lines:

**Case 1**
```
@discriminator("kind")
model Pet<TDiscriminator, TParams> {
  kind: TDiscriminator;
  parameters: TParams;
}

model Cat extends Pet<"Cat", CatParams> {}
model CatParams {
  lives: int32;
}

model Dog extends Pet<"Dog", DogParams> {}
model DogParams {
  ageInDogYears: int32;
}

interface PetStore {
  @route(":adopt")
  @post adoptPet(@body body: Pet): Response
}
```
This made total intuitive sense to me, but I was dismayed to find this didn't work because it says I didn't specify enough template parameters.

After some discussion, I found that this, _did_ work:

**Case 2**
```
@discriminator("kind")
model Pet {}
model PetTemplate<TDiscriminator, TParams> extends Pet {
  kind: TDiscriminator;
  parameters: TParams;
}

model Cat extends PetTemplate<"Cat", CatParams> {}
model CatParams {
  lives: int32;
}

model Dog extends PetTemplate<"Dog", DogParams> {}
model DogParams {
  ageInDogYears: int32;
}

interface PetStore {
  @route(":adopt")
  @post adoptPet(@body body: Pet): Response
}
```
The fact that this worked... left my annoyed, confused and unhappy. That I'm feeding the operation a type that is basically an empty object makes no sense. The `@discriminator` decorator doesn't add the kind property to Pet, so if there were no Cat or Dog models, the CADL compiler is happy about a type that is inherently nonsensical. It promises a property that it clearly doesn't have, essentially creating some kind of abstract base class. However, if that is fine, then the compiler should also be happy with me simply providing a template name to serve the same role. 

**Expected Behavior**

As a service author, I would simply expect that if I put a template as a parameter or response, it would implicitly be interpreted as "any model which specializes this template". In practical terms, for this scenario it would effectively make the interface function like:
```
interface PetStore {
  @route(":adopt")
  @post adoptPet(@body body: Cat | Dog): Response
}
```

If there are deeper philosophical issues about accepting generic templates in operation signatures, then perhaps consider having the compiler treat this like syntactic sugar. Rather than fully support templates as generic types, simply have it convert instances of naked template names into "the union of all models which extend a specialized template". Essentially an implicit:

`alias Pet = Cat | Dog;`

This avoids having to update multiple places when a new Pet is added. While I could certainly create the alias manually, if the CADL is split across multiple files (which they tend to be) then it means to add `Bird` I will need to update things in multiple places. With this syntactic sugar, I would merely need to ensure that `Bird` extends `Pet`.

**Comment by bterlson** [comment]
While I try not to beat this drum so much... I think the primary issue here is the way we are defining discriminated unions using `extends` which as far as cadl is concerned is just a subclassing relationship and nothing more. We could make `@discriminator` on a model change the semantics of the model such that references to the base class is sugar for a union of its subclasses and make the rest of what you want work, but at that point we should just use a union declaration IMO. I continue to believe that #335 is the best approach, with an optional `extends` to specify the constraint that all union variants must extend a common base type.

The example above with #335 and extends would look like:

```
model PetBase<TDiscriminator, TParams> {
  kind: TDiscriminator;
  parameters: TParams;
}

model Cat extends PetBase<"Cat", CatParams> {}
model CatParams {
  lives: int32;
}

model Dog extends PetBase<"Dog", DogParams> {}
model DogParams {
  ageInDogYears: int32;
}

@discriminator("kind")
union Pet extends PetBase { Cat, Dog }

interface PetStore {
  @route(":adopt")
  @post adoptPet(@body body: Pet): Response
}
```

In general I would say the notion that an uninstantiated template can be substituted for any of its instantiations or subclasses thereof is not type safe in the sense that whether the substitution is valid depends on the variance of the type parameters and, in cadl, what decorators are doing with the type parameters. It also seems quite likely that the user just forgot the template parameter, and so implicitly engaging some behavior which will do a thing seems risky (even if gated on the model carrying the @discriminator decorator). I believe that cadl type references should continue to require instantiation.

**Comment by markcowl** [comment]
@bterlson Is the 'Petbase' referenced in this union: `union Pet extends PetBase { Cat, Dog }` missing from the above snippet, or is it somehow derived from the leaf types?  Generally, being explicit about the definition of the base type is desirable, I think, especially in cases where the base type contains additional proeprties beyond the discriminator.

**Comment by bterlson** [comment]
It's on line 1 of the snippet (if I understand the question  correctly).

**Comment by markcowl** [comment]
closing as duplicate of #335
